test/e2e/framework/ginkgowrapper.go

```diff
@@ -67,6 +125,322 @@ func AnnotatedLocationWithOffset(annotation string, offset int) types.CodeLocati
+// of the text arguments and special tags from the With functions.
+func It(text string, args ...interface{}) bool {
+	return registerInSuite(ginkgo.It, text, args)
+}
+
+// It is a shorthand for the corresponding package function.
+func (f *Framework) It(text string, args ...interface{}) bool {
+	return registerInSuite(ginkgo.It, text, args)
+}
+
+// Describe is a wrapper around [ginkgo.Describe] which supports framework
+// With* labels as optional arguments in addition to those already supported by
+// ginkgo itself, like [ginkgo.Label] and [gingko.Offset].
+//
+// Text and arguments may be mixed. The final text is a concatenation
+// of the text arguments and special tags from the With functions.
+func Describe(text string, args ...interface{}) bool {
+	return registerInSuite(ginkgo.Describe, text, args)
+}
+
+// Describe is a shorthand for the corresponding package function.
+func (f *Framework) Describe(text string, args ...interface{}) bool {
+	return registerInSuite(ginkgo.Describe, text, args)
+}
+
+// Context is a wrapper around [ginkgo.Context] which supports framework With*
+// labels as optional arguments in addition to those already supported by
+// ginkgo itself, like [ginkgo.Label] and [gingko.Offset].
+//
+// Text and arguments may be mixed. The final text is a concatenation
+// of the text arguments and special tags from the With functions.
+func Context(text string, args ...interface{}) bool {
+	return registerInSuite(ginkgo.Context, text, args)
+}
+
+// Context is a shorthand for the corresponding package function.
+func (f *Framework) Context(text string, args ...interface{}) bool {
+	return registerInSuite(ginkgo.Context, text, args)
+}
+
+// registerInSuite is the common implementation of all wrapper functions. It
+// expects to be called through one intermediate wrapper.
+func registerInSuite(ginkgoCall func(text string, args ...interface{}) bool, text string, args []interface{}) bool {
+	var ginkgoArgs []interface{}
+	var offset ginkgo.Offset
+	var texts []string
+	if text != "" {
+		texts = append(texts, text)
+	}
+
+	addLabel := func(label string) {
+		texts = append(texts, fmt.Sprintf("[%s]", label))
+		ginkgoArgs = append(ginkgoArgs, ginkgo.Label(label))
+	}
+
+	haveEmptyStrings := false
+	for _, arg := range args {
+		switch arg := arg.(type) {
+		case label:
+			fullLabel := strings.Join(arg.parts, ": ")
+			addLabel(fullLabel)
+			if arg.extra != "" {
+				addLabel(arg.extra)
+			}
+			if fullLabel == "Serial" {
+				ginkgoArgs = append(ginkgoArgs, ginkgo.Serial)
+			}
+		case ginkgo.Offset:
+			offset = arg
+		case string:
+			if arg == "" {
+				haveEmptyStrings = true
+			}
+			texts = append(texts, arg)
+		default:
+			ginkgoArgs = append(ginkgoArgs, arg)
+		}
+	}
+	offset += 2 // This function and its direct caller.
+
+	// Now that we have the final offset, we can record bugs.
+	if haveEmptyStrings {
+		RecordBug(NewBug("empty strings as separators are unnecessary and need to be removed", int(offset)))
```

Comment:
what's the deal with the spaces and empty strings, I can't remember

Reply:
The final `strings.Join(" ", components)` will add two spaces around any empty string - just a nit, but not nice.